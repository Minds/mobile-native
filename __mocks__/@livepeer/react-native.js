// mock @livepeer/react-native
export const LivepeerConfig = {
  setConfig: jest.fn(),
  getConfig: jest.fn(),
  get: jest.fn(),
  set: jest.fn(),
};

export const createReactClient = jest.fn();
export const createLivepeerClient = jest.fn();
export const Player = jest.fn();
