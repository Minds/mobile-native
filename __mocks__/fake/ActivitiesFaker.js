function load(count, container) {
  let activities = [...Array(count)].map((_, i) => {
    const code = 'activityguid' + i;
    return {
      attachment_guid: false,
      blurb: false,
      container_guid: code,
      custom_data: false,
      custom_type: false,
      rowKey: 'something' + i,
      description: 'Congratulations! ',
      edited: '',
      guid: code,
      mature: false,
      time_created: '1522036284',
      ownerObj: {
        guid: '824853017709780997',
        type: 'user',
        subtype: false,
        time_created: '1522036284',
      },
      message: 'Message',
      title: 'TITLE',
      owner_guid: '824853017709780997',
      parent_guid: '838106762591510528',
      perma_url: false,
      thumbnail_src: false,
      type: 'activity',
      wire_totals: {
        tokens: 1000000000000000000,
      },
      containerObj: container,
    };
  });

  return {
    activities: activities,
    'load-next': 'aaaaaa',
    'load-previous': 'aaaaaa',
  };
}

/**
 * comment service data faker
 * @param {integer} count
 */
export function activitiesServiceFaker() {
  return { load: load };
}

/**
 * Generate a single fake entity
 * @param {object} containerObj optional entity container
 * @returns
 */
export function fakeOne(containerObj) {
  const resp = load(1, containerObj);
  return resp.activities[0];
}
