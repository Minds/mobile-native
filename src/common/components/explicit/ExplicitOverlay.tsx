import React, { PureComponent } from 'react';

import {
  Text,
  StyleSheet,
  View,
  TextStyle,
  ViewStyle,
  TouchableOpacity,
} from 'react-native';

import Icon from 'react-native-vector-icons/SimpleLineIcons';
import { GOOGLE_PLAY_STORE } from '../../../config/Config';
import i18n from '../../services/i18n.service';
import type ActivityModel from '../../../newsfeed/ActivityModel';
import type UserModel from '../../../channel/UserModel';
import ThemedStyles from '../../../styles/ThemedStyles';
import GroupModel from '../../../groups/GroupModel';

type PropsType = {
  entity: ActivityModel | UserModel | GroupModel;
  iconSize: number;
  hideText: boolean;
  fontStyle?: TextStyle | Array<TextStyle>;
  containerStyle?: ViewStyle | Array<ViewStyle>;
  text?: string;
};

/**
 * Explicit overlay
 */
export default class ExplicitOverlay extends PureComponent<PropsType> {
  /**
   * Default props
   */
  static defaultProps = {
    iconSize: 55,
    hideText: false,
    iconPosition: 'right',
  };

  /**
   * toggle overlay
   */
  toggle = () => {
    this.props.entity.toggleMatureVisibility();
  };

  getLocalizedReasons() {
    const reasons = this.props.entity.nsfw?.map((i) => i18n.t(`nsfw.${i}`));

    if (!reasons) {
      return '';
    }

    if (reasons.length === 1) {
      return reasons[0];
    }
    if (reasons.length === 2) {
      return reasons.join(' & ');
    }
    if (reasons.length > 2) {
      return (
        reasons.slice(0, reasons.length - 1).join(', ') +
        ' & ' +
        reasons[reasons.length - 1]
      );
    }
  }

  /**
   * Render
   */
  render() {
    const { iconSize, hideText, fontStyle, containerStyle } = this.props;

    const theme = ThemedStyles.style;

    if (this.props.entity.mature_visibility) {
      return null;
    }

    const text = GOOGLE_PLAY_STORE
      ? i18n.t('postCantBeShown')
      : i18n.t('confirm18');

    return (
      <View
        style={[
          theme.centered,
          theme.backgroundSecondary,
          styles.onTop,
          containerStyle,
        ]}>
        <Icon
          name="lock"
          size={iconSize}
          style={theme.colorSecondaryText}
          onPress={this.toggle}
        />
        {!hideText && (
          <>
            {!!this.props.text && (
              <Text
                style={[
                  theme.colorPrimaryText,
                  theme.paddingTop6x,
                  theme.fontXL,
                  theme.bold,
                  fontStyle,
                ]}>
                {this.props.text}
              </Text>
            )}
            <Text
              style={[
                theme.colorPrimaryText,
                theme.paddingTop6x,
                theme.fontXXL,
                theme.bold,
              ]}>
              NSFW
            </Text>
            <Text
              style={[
                theme.colorTertiaryText,
                theme.paddingVertical2x,
                theme.marginBottom6x,
                theme.fontXL,
              ]}>
              {this.getLocalizedReasons()}
            </Text>
            <TouchableOpacity
              style={[
                theme.borderPrimary,
                theme.borderHair,
                theme.backgroundPrimary,
              ]}
              onPress={this.toggle}>
              <Text
                style={[theme.col, theme.padding2x, theme.fontL, fontStyle]}>
                {text}
              </Text>
            </TouchableOpacity>
          </>
        )}
      </View>
    );
  }
}

const styles = StyleSheet.create({
  onTop: {
    minHeight: 400,
  },
});
