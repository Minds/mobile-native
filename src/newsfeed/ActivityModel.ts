import { runInAction, action, observable, decorate } from 'mobx';
import FastImage from 'react-native-fast-image';
import { FlatList, Alert, Platform } from 'react-native';

import BaseModel from '../common/BaseModel';
import UserModel from '../channel/UserModel';
import wireService from '../wire/WireService';
import sessionService from '../common/services/session.service';
import {
  setPinPost,
  deleteItem,
  unfollow,
  follow,
} from '../newsfeed/NewsfeedService';
import api from '../common/services/api.service';

import { GOOGLE_PLAY_STORE, MINDS_CDN_URI, MINDS_URI } from '../config/Config';
import i18n from '../common/services/i18n.service';
import logService from '../common/services/log.service';
import type { ThumbSize, LockType } from '../types/Common';
import type GroupModel from '../groups/GroupModel';
import { SupportTiersType } from '../wire/WireTypes';
import mindsService from '../common/services/minds.service';
import NavigationService from '../navigation/NavigationService';
import { showNotification } from '../../AppMessages';

type Thumbs = Record<ThumbSize, string> | Record<ThumbSize, string>[];

/**
 * Activity model
 */
export default class ActivityModel extends BaseModel {
  // Observable properties
  @observable pinned: boolean = false;
  @observable time_created: string = '';
  @observable message: string = '';
  @observable title: string = '';
  @observable mature: boolean = false;
  @observable edited: '0' | '1' = '0';
  @observable paywall: true | '1' | '' = '';

  // decorated observables
  'is:following': boolean;
  'thumbs:down:count': number;
  'thumbs:up:count': number;
  'comments:count': number;
  'thumbs:down:user_guids': Array<number>;
  'thumbs:up:user_guids': Array<number>;
  seen?: boolean;
  rowKey?: string;
  description?: string; // on image objects in some cases the message is on description field
  containerObj?: GroupModel;
  remind_object?: ActivityModel;
  ownerObj!: UserModel;
  listRef?: FlatList<any>;
  thumbnails?: Thumbs;
  paywall_unlocked: boolean = false;
  guid: string = '';
  subtype: string = '';
  entity_guid: string = '';
  owner_guid: string = '';
  custom_type?: string;
  custom_data?: Array<any> | any;
  nsfw?: Array<number>;
  flags?: any;
  reminds: number = 0;
  impressions: number = 0;
  perma_url?: string;
  cinemr_guid?: string;
  thumbnail_src?: string;
  dontPin?: boolean;
  boosted?: boolean;
  wire_threshold?: {
    type: 'token' | 'money';
    min: number;
    support_tier: SupportTiersType;
  } | null;
  _preview?: boolean;
  attachments?: {
    attachment_guid: string;
  };
  type?: string;
  permaweb_id?: string;
  remind_deleted?: boolean;
  remind_users?: Array<UserModel>;

  /**
   * Mature visibility flag
   */
  @observable mature_visibility: boolean = false;

  /**
   * Is visible in flat list
   */
  @observable is_visible: boolean = false;

  /**
   *  List reference setter
   */
  set _list(value) {
    this.__list = value;

    // the reminded object need access to the metadata service too
    if (this.remind_object) {
      this.remind_object._list = value;
    }
  }

  /**
   *  List reference getter
   */
  get _list() {
    return this.__list;
  }

  /**
   * Block owner
   */
  async blockOwner() {
    await super.blockOwner();
    if (this._list) {
      this._list.refresh();
    }
  }

  /**
   * Unblock owner
   */
  async unblockOwner() {
    await super.unblockOwner();
    if (this._list) {
      this._list.refresh();
    }
  }

  /**
   * Has media
   */
  hasMedia(): boolean {
    const type = this.custom_type || this.subtype;
    switch (type) {
      case 'image':
      case 'video':
      case 'batch':
        return true;
    }
    if (this.perma_url) {
      return true;
    }
    return false;
  }

  /**
   * Has an attached video
   */
  hasVideo(): boolean {
    return (this.custom_type || this.subtype) === 'video';
  }

  /**
   * Has an attached image
   */
  hasImage(): boolean {
    const type = this.custom_type || this.subtype;
    return type === 'image' || type === 'batch';
  }

  hasThumbnails(): boolean {
    return Array.isArray(this.thumbnails) && this.thumbnails.length === 0
      ? false
      : this.thumbnails
      ? true
      : false;
  }

  toPlainObject() {
    const plainEntity = super.toPlainObject();
    if (plainEntity.remind_object && plainEntity.remind_object.__list) {
      delete plainEntity.remind_object.__list;
    }

    return plainEntity;
  }

  /**
   * Child models
   */
  childModels(): any {
    return {
      ownerObj: UserModel,
      remind_object: ActivityModel,
    };
  }

  /**
   * Get the activity thumb source
   * {uri: 'http...'}
   * @param {string} size
   */
  getThumbSource(size: ThumbSize = 'medium') {
    // for gif use always the same size to take advantage of the cache (they are not resized)
    if (this.isGif()) {
      size = 'xlarge';
    }

    if (this.thumbnails && this.thumbnails[size]) {
      return { uri: this.thumbnails[size], headers: api.buildHeaders() };
    }

    // fallback to old behavior
    if (this.paywall || this.paywall_unlocked) {
      const guid = this.entity_guid === '' ? this.guid : this.entity_guid;
      const unlock = this.isOwner() ? '?unlock_paywall=1' : '';
      return {
        uri: MINDS_URI + 'fs/v1/thumbnail/' + guid + unlock,
        headers: api.buildHeaders(),
      };
    }
    if (this.custom_type === 'batch') {
      return {
        uri: MINDS_CDN_URI + 'fs/v1/thumbnail/' + this.entity_guid + '/' + size,
        headers: api.buildHeaders(),
      };
    }
    return {
      uri: MINDS_CDN_URI + 'fs/v1/thumbnail/' + this.guid + '/' + size,
      headers: api.buildHeaders(),
    };
  }

  isGif(): boolean {
    return this.custom_data && this.custom_data[0] && this.custom_data[0].gif;
  }

  /**
   * Preload thumb on image cache
   */
  preloadThumb(size: ThumbSize = 'medium') {
    FastImage.preload([this.getThumbSource(size)]);
  }

  shouldBeBlured(): boolean {
    const user = sessionService.getUser();

    if (user && user.mature) {
      return false;
    }

    if (this.nsfw !== undefined) {
      let res = [1, 2, 4].filter((nsfw) => this.nsfw!.indexOf(nsfw) > -1)
        .length;
      if (res) {
        return true;
      }
    }

    if (typeof this.flags !== 'undefined') {
      return !!this.flags.mature;
    }

    if (typeof this.mature !== 'undefined') {
      return !!this.mature;
    }

    if (
      typeof this.custom_data !== 'undefined' &&
      typeof this.custom_data[0] !== 'undefined'
    ) {
      return !!this.custom_data[0].mature;
    }

    if (this.custom_data !== undefined) {
      return !!this.custom_data.mature;
    }

    return false;
  }

  /**
   * Get activity text
   */
  get text() {
    return this.message || this.description || '';
  }

  @action
  toggleMatureVisibility = () => {
    if (GOOGLE_PLAY_STORE || Platform.OS === 'ios') {
      showNotification(i18n.t('activity.notSafeComment'));
      return;
    }
    this.mature_visibility = !this.mature_visibility;

    if (this.remind_object && this.remind_object) {
      this.remind_object.mature_visibility = this.mature_visibility;
    }
  };

  @action
  setVisible(value: boolean) {
    this.is_visible = value;
    if (this.remind_object) {
      this.remind_object.is_visible = value;
    }
  }

  /**
   * Increment the comments counter
   */
  @action
  incrementCommentsCounter() {
    this['comments:count']++;
  }

  /**
   * Decrement the comments counter
   */
  @action
  decrementCommentsCounter() {
    this['comments:count']--;
  }

  /**
   * Unlock the activity and update data on success
   */
  @action
  async unlock(ignoreError = false) {
    try {
      const result: object | false = await wireService.unlock(this.guid);
      if (result) {
        // all changes should be atomic (trigger render only once)
        runInAction(() => {
          // create a new model because we need the child models
          const list = this.__list;
          const model: ActivityModel = ActivityModel.create(result);
          Object.assign(this, model);
          this.__list = list;
        });
      }

      if (
        this.entity_guid &&
        this.perma_url &&
        this.perma_url?.startsWith(MINDS_URI)
      ) {
        NavigationService.push('BlogView', {
          guid: this.entity_guid,
          unlock: true,
        });
      }

      return result;
    } catch (err) {
      if (!ignoreError) {
        Alert.alert(err.message);
      }
      return false;
    }
  }

  /**
   * Get the lock type for the activity
   */
  getLockType = (): LockType | undefined => {
    const support_tier: SupportTiersType | null =
      this.wire_threshold && 'support_tier' in this.wire_threshold
        ? this.wire_threshold.support_tier
        : null;
    if (!support_tier) {
      return;
    }
    let type: LockType = support_tier.public ? 'members' : 'paywall';

    if (mindsService.settings.plus.support_tier_urn === support_tier.urn) {
      type = 'plus';
    }

    return type;
  };

  /**
   * Unlock the entity or prompt pay options
   */
  async unlockOrPay() {
    const result = await this.unlock(true);

    if (result) {
      return;
    }

    const lockType = this.getLockType();

    const support_tier: SupportTiersType | null =
      this.wire_threshold && 'support_tier' in this.wire_threshold
        ? this.wire_threshold.support_tier
        : null;

    switch (lockType) {
      case 'plus':
        NavigationService.push('UpgradeScreen', {
          support_tier,
          entity: this,
          onComplete: (resultComplete: any) => {
            if (resultComplete && resultComplete.payload.method === 'onchain') {
              setTimeout(() => {
                Alert.alert(
                  i18n.t('wire.weHaveReceivedYourTransaction'),
                  i18n.t('wire.pleaseTryUnlockingMessage'),
                );
              }, 400);
            } else {
              this.unlock();
            }
          },
        });
        break;
      case 'members':
      case 'paywall':
        NavigationService.push('JoinMembershipScreen', {
          entity: this,
          onComplete: (resultComplete: any) => {
            if (resultComplete && resultComplete.payload.method === 'onchain') {
              setTimeout(() => {
                Alert.alert(
                  i18n.t('wire.weHaveReceivedYourTransaction'),
                  i18n.t('wire.pleaseTryUnlockingMessage'),
                );
              }, 400);
            } else {
              this.unlock();
            }
          },
        });
    }
  }

  @action
  async togglePin() {
    // allow owners only
    if (!this.isOwner()) {
      return;
    }

    try {
      this.pinned = !this.pinned;
      await setPinPost(this.guid, this.pinned);
    } catch (e) {
      this.pinned = !this.pinned;
      Alert.alert(i18n.t('errorPinnedPost'));
    }
  }

  @action
  async deleteEntity() {
    try {
      await deleteItem(this.guid);
      this.removeFromList();
      ActivityModel.events.emit('deleteEntity', this);
    } catch (err) {
      logService.exception('[ActivityModel]', err);
      throw err;
    }
  }

  async deleteRemind() {
    try {
      await api.delete(`api/v3/newsfeed/${this.urn}`);
      this.removeFromList();
      ActivityModel.events.emit('deleteEntity', this);
    } catch (err) {
      logService.exception('[ActivityModel]', err);
      throw err;
    }
  }

  @action
  async toggleFollow() {
    const method = this['is:following'] ? unfollow : follow;
    try {
      await method(this.guid);
      runInAction(() => {
        this['is:following'] = !this['is:following'];
      });
    } catch (err) {
      logService.exception('[OffsetFeedListStore]', err);
      throw err;
    }
  }

  @action
  setEdited(value) {
    this.edited = value;
  }
}

/**
 * Define model observables
 */
decorate(ActivityModel, {
  //@ts-ignore
  'thumbs:down:count': observable,
  'thumbs:up:count': observable,
  'comments:count': observable,
  'thumbs:down:user_guids': observable,
  'thumbs:up:user_guids': observable,
});
