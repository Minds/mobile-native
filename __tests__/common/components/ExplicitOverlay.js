import 'react-native';
import React from 'react';
import { shallow } from 'enzyme';
import ExplicitOverlay from '~/common/components/explicit/ExplicitOverlay';
import ActivityModel from '~/newsfeed/ActivityModel';
import { activitiesServiceFaker } from '../../../__mocks__/fake/ActivitiesFaker';
import sp from '~/services/serviceProvider';

jest.mock('~/services/serviceProvider');

// mock services
sp.mockService('styles');
sp.mockService('i18n');

describe('Explicit overlay component', () => {
  let entity, screen;
  beforeEach(() => {
    let mockResponse = activitiesServiceFaker().load(1);
    entity = ActivityModel.create(mockResponse.activities[0]);
    entity.mature_visibility = false;
    screen = shallow(<ExplicitOverlay entity={entity} />);
  });

  it('renders correctly', async () => {
    expect(screen).toMatchSnapshot();
  });

  it('should renders the etag', async () => {
    entity.mature_visibility = true;
    await screen.update();

    expect(screen).toMatchSnapshot();
  });
});
