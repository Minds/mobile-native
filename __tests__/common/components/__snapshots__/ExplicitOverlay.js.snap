// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Explicit overlay component renders correctly 1`] = `
<View
  style={
    Array [
      undefined,
      undefined,
      Object {
        "minHeight": 400,
      },
      undefined,
    ]
  }
>
  <Icon
    allowFontScaling={false}
    name="lock"
    onPress={[Function]}
    size={55}
  />
  <Text
    style={
      Array [
        undefined,
        undefined,
        undefined,
        undefined,
      ]
    }
  >
    NSFW
  </Text>
  <Text
    style={
      Array [
        undefined,
        undefined,
        undefined,
        undefined,
      ]
    }
  />
  <TouchableOpacity
    onPress={[Function]}
    style={
      Array [
        undefined,
        undefined,
        undefined,
      ]
    }
  >
    <Text
      style={
        Array [
          undefined,
          undefined,
          undefined,
          undefined,
        ]
      }
    >
      I'm over 18 years of age
    </Text>
  </TouchableOpacity>
</View>
`;

exports[`Explicit overlay component should renders the etag 1`] = `
<View
  style={
    Array [
      undefined,
      undefined,
      Object {
        "minHeight": 400,
      },
      undefined,
    ]
  }
>
  <Icon
    allowFontScaling={false}
    name="lock"
    onPress={[Function]}
    size={55}
  />
  <Text
    style={
      Array [
        undefined,
        undefined,
        undefined,
        undefined,
      ]
    }
  >
    NSFW
  </Text>
  <Text
    style={
      Array [
        undefined,
        undefined,
        undefined,
        undefined,
      ]
    }
  />
  <TouchableOpacity
    onPress={[Function]}
    style={
      Array [
        undefined,
        undefined,
        undefined,
      ]
    }
  >
    <Text
      style={
        Array [
          undefined,
          undefined,
          undefined,
          undefined,
        ]
      }
    >
      I'm over 18 years of age
    </Text>
  </TouchableOpacity>
</View>
`;
